kornia/feature/mkd.py

```diff
@@ -0,0 +1,609 @@
+
+        # Polar embedding with relative gradients.
+        if dtype in ['polar', 'concat']:
+            ori_rel = EmbedGradients(patch_size=patch_size,
+                                     relative=True)
+            polar_emb = ExplicitSpacialEncoding(dtype='polar',
+                                                fmap_size=patch_size,
+                                                in_dims=ori_rel.kernel.d)
+            self.polar_feats = nn.Sequential(ori_rel, polar_emb)
+
+        self.odims: int = 0
+        if dtype == 'concat':
+            self.odims = polar_emb.odims + cart_emb.odims
+        elif dtype == 'cart':
+            self.odims = cart_emb.odims
+        elif dtype == 'polar':
+            self.odims = polar_emb.odims
+
+        # Compute true reduce_dims.
+        self.reduce_dims: int = min(reduce_dims, self.odims)
+
+        # Load supervised(lw)/unsupervised(pca) model trained on training_set.
+        if self.whitening is not None:
+            whitening_models = torch.hub.load_state_dict_from_url(
+                urls[self.dtype], map_location=lambda storage, loc: storage
+            )
+            whitening_model = whitening_models[training_set]
+            self.whitening_layer = Whitening(whitening,
+                                             whitening_model,
+                                             in_dims=self.odims,
+                                             reduce_dims=self.reduce_dims)
+            self.odims = self.reduce_dims
+
+    def forward(self, patches: torch.Tensor) -> torch.Tensor:
+        # Extract gradients.
+        g = self.smoothing(patches)
+        g = self.gradients(g)
+
+        # Extract polar and/or cart features.
+        if self.dtype in ['polar', 'concat']:
+            pe = self.polar_feats(g)
+        if self.dtype in ['cart', 'concat']:
+            ce = self.cart_feats(g)
+
+        # Concatenate.
+        if self.dtype == 'concat':
+            y = torch.cat([pe, ce], dim=1)
+        elif self.dtype == 'cart':
+            y = ce
+        elif self.dtype == 'polar':
+            y = pe
+
+        # l2-normalize.
+        y = F.normalize(y, dim=1)
+
+        # Whiten descriptors.
+        if self.whitening is not None:
+            y = self.whitening_layer(y)
+
+        return y
+
+    def __repr__(self) -> str:
+        return self.__class__.__name__ +\
+            '(' + 'patch_size=' + str(self.patch_size) +\
+            ', ' + 'dtype=' + str(self.dtype) +\
+            ', ' + 'whitening=' + str(self.whitening) +\
+            ', ' + 'training_set=' + str(self.training_set) +\
+            ', ' + 'reduce_dims=' + str(self.reduce_dims) + ')'
+
+
+def load_whitening_model(dtype: str, training_set: str) -> Dict:
+    whitening_models = torch.hub.load_state_dict_from_url(
+        urls[dtype], map_location=lambda storage, loc: storage
+    )
+    whitening_model = whitening_models[training_set]
+    return whitening_model
+
+
+class SimpleKD(nn.Module):
```

Comment:
is this exposed to the highlevel api too ?

Reply:
No, it was purely an example to illustrate that without 'concat', the whole thing can be written as a sequence.

Reply:
ok then I suggest that everything that won't go to the high level API goes with underscore in front. This includes classes too